analyzers/src/SonarAnalyzer.CSharp/Rules/CollectionEmptinessChecking.cs

```diff
@@ -36,115 +36,70 @@ public sealed class CollectionEmptinessChecking : SonarDiagnosticAnalyzer
 
+        private static readonly CSharpExpressionNumericConverter ExpressionNumericConverter = new CSharpExpressionNumericConverter();
+
         public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(rule);
 
-        protected override void Initialize(SonarAnalysisContext context)
-        {
-            context.RegisterSyntaxNodeActionInNonGenerated(
-                c =>
-                {
-                    var binary = (BinaryExpressionSyntax) c.Node;
-                    CheckCountZero(binary.Right, binary.Left, c);
-                    CheckCountOne(binary.Left, binary.Right, c);
-                },
-                SyntaxKind.GreaterThanExpression);
-            context.RegisterSyntaxNodeActionInNonGenerated(
-                c =>
-                {
-                    var binary = (BinaryExpressionSyntax)c.Node;
-                    CheckCountZero(binary.Left, binary.Right, c);
-                    CheckCountOne(binary.Right, binary.Left, c);
-                },
-                SyntaxKind.LessThanExpression);
-            context.RegisterSyntaxNodeActionInNonGenerated(
-                c =>
-                {
-                    var binary = (BinaryExpressionSyntax)c.Node;
-                    CheckCountOne(binary.Right, binary.Left, c);
-                    CheckCountZero(binary.Left, binary.Right, c);
-                },
-                SyntaxKind.GreaterThanOrEqualExpression);
-            context.RegisterSyntaxNodeActionInNonGenerated(
-                c =>
-                {
-                    var binary = (BinaryExpressionSyntax)c.Node;
-                    CheckCountOne(binary.Left, binary.Right, c);
-                    CheckCountZero(binary.Right, binary.Left, c);
-                },
-                SyntaxKind.LessThanOrEqualExpression);
-            context.RegisterSyntaxNodeActionInNonGenerated(
-                c =>
-                {
-                    var binary = (BinaryExpressionSyntax)c.Node;
-                    CheckCountZero(binary.Left, binary.Right, c);
-                    CheckCountZero(binary.Right, binary.Left, c);
-                },
-                SyntaxKind.EqualsExpression);
+        protected override void Initialize(SonarAnalysisContext context) =>
             context.RegisterSyntaxNodeActionInNonGenerated(
                 c =>
                 {
                     var binary = (BinaryExpressionSyntax)c.Node;
-                    CheckCountZero(binary.Left, binary.Right, c);
-                    CheckCountZero(binary.Right, binary.Left, c);
+                    var left = ExpressionNumericConverter.TryGetConstantIntValue(binary.Left, out var l_out) ? (int?)l_out : null;
+                    var right = ExpressionNumericConverter.TryGetConstantIntValue(binary.Right, out var r_out) ? (int?)r_out : null;
+
+                    if ((left ?? right) is int constant)
```

Comment:
Can we pre-filter expected constant values to save some code execution on a quick-win condition? 
```suggestion
                    if ((left ?? right) is int constant && (constant == 0 || constant == 1))
```

Reply:
Technically we can. But I would argue that the case will not be easier to understand, and is not worth the (premature) optimization.